// app/map/index.js
import proj4 from 'proj4/lib';
import axios from 'axios';
import ox from 'ox';

import * as config from '../config';

let _olMap = null;

export function get() {
    if (_olMap)
        return _olMap;

    _olMap = new ox.Map({
        controls: [],
        layers: [],
    });

    return _olMap;
}

export async function init() {
    ox.proj.setProj4(proj4);

    let p = config.str('map.proj.server');
    proj4.defs(p, config.str('map.proj.defs.' + p));

    // let res = await axios.get(config.str('server.wms.url'), {params: {
    //     'SERVICE': 'WMS',
    //     'VERSION': '1.3',
    //     'REQUEST': 'GetProjectSettings'
    //
    // }});
    //
    // let wmsCaps = new ol.format.WMSCapabilities().read(res.data);
    //
    // console.log(wmsCaps)



    let layers = [
        new ox.layer.Tile({
            source: new ox.source.OSM(),
        }),
        new ox.layer.Image({
            source: new ox.source.ImageWMS({
                url: config.str('server.wms.url'),
                projection: config.str('map.proj.server'),
                params: {
                    LAYERS: config.str('server.wms.layers')
                },
                ratio: 1,
                serverType: 'geoserver'
            })
        }),
    ];

    get().addLayer(layers[0]);
    get().addLayer(layers[1]);
    get().setView(new ox.View({
        center: ox.proj.fromLonLat([9.9867157, 53.5414746]),
        zoom: 18
    }));

    /*global Promise*/
    return Promise.resolve(get());

}
