// app/map/index.ts
import * as ol from 'openlayers';
import proj4 from 'proj4';
import axios from 'axios';

import * as config from '../config';


let _olMap = null;

function get(): ol.Map {
    if (_olMap)
        return _olMap;

    _olMap = new ol.Map({
        controls: [],
        layers: [],
    });

    return _olMap;
}

async function init(): Promise<ol.Map> {
    ol.proj.setProj4(proj4);

    let p = config.str('map.proj.server');
    proj4.defs(p, config.str('map.proj.defs.' + p));

    let res = await axios.get(config.str('server.wms.url'), {params: {
        'SERVICE': 'WMS',
        'VERSION': '1.3',
        'REQUEST': 'GetProjectSettings'

    }});

    let wmsCaps = new ol.format.WMSCapabilities().read(res.data);

    console.log(wmsCaps)



    let layers = [
        new ol.layer.Tile({
            source: new ol.source.OSM(),
        }),
        new ol.layer.Image({
            source: new ol.source.ImageWMS({
                url: config.str('server.wms.url'),
                projection: config.str('map.proj.server'),
                params: {
                    LAYERS: config.str('server.wms.layers')
                },
                ratio: 1,
                serverType: 'geoserver'
            })
        }),
    ];

    get().addLayer(layers[0]);
    get().addLayer(layers[1]);
    get().setView(new ol.View({
        center: ol.proj.fromLonLat([9.9867157, 53.5414746]),
        zoom: 18
    }));

    return Promise.resolve(get());

}


export {get, init};
