// components/ColorPicker.js
import React from 'react';
import Popover from 'material-ui/Popover';

export default class ColorPicker extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            open: false,
        };
    }

    openClick(evt) {
        evt.preventDefault();
        this.setState({
            open: true,
            anchorEl: evt.currentTarget,
        });
    }

    requestClose() {
        this.setState({
            open: false,
        });
    }

    valueClick(evt, color) {
        this.setState({
            open: false
        }, () => this.props.onChange(evt, color));
    }

    render() {
        let cstyle = {
            width: 32,
            height: 32,
            margin: 8,
            cursor: 'pointer',
            borderRadius: '50%'
        };

        let val = this.props.value || '#aaa',
            width = this.props.width || 200;

        return (
            <div>
                <div
                    onClick={e => this.openClick(e)}
                    style={{backgroundColor: val, ...cstyle}}
                />
                <Popover
                    open={this.state.open}
                    anchorEl={this.state.anchorEl}
                    onRequestClose={() => this.requestClose()}
                    style={{width}}
                >{this.props.colors.map(c => <div
                        key={c}
                        style={{backgroundColor: c, float: 'left', ...cstyle}}
                        onClick={evt => this.valueClick(evt, c)}
                    />
                )}</Popover>
            </div>

        )
    }
}
