// components/IFrame.js
import React from 'react';
import Paper from 'material-ui/Paper';
import MaterialIcon from './MaterialIcon';
import IconButton from 'material-ui/IconButton';
import {Toolbar, ToolbarGroup, ToolbarSeparator, ToolbarTitle} from 'material-ui/Toolbar';
import withWidth, {SMALL} from 'material-ui/utils/withWidth';
import muiThemeable from 'material-ui/styles/muiThemeable';

class IFrame extends React.Component {
    render() {
        let mobile = this.props.width === SMALL,
            style = {
            paper : {
                position : 'fixed',
                top: mobile ? 0 : '10vh',
                right: mobile ? 0 : '15vw', 
                width : mobile ? '100vw' : '70vw',
                maxWidth : mobile ? '100vw' : '70vw',
                height : mobile ? '100vh' : '80vh',
                maxHeight : mobile ? '100vh' : '80vh',
                visibility : this.props.open ? 'visible' : 'hidden',
            },
            iframe : {
                width: '100%',
                height: 'calc(100% - ' + this.props.muiTheme.appBar.height + 'px)',
                border: 'none',
            }
        };
        return (
            <div>
                <Paper
                    style={style.paper}
                >
                    <Toolbar>
                        <ToolbarGroup>
                            <ToolbarTitle text={this.props.title} />
                        </ToolbarGroup>
                        <ToolbarGroup lastChild={true}>
                            <IconButton
                                style={style.icon}
                                onClick={this.props.onRequestClose}
                            >
                                <MaterialIcon
                                    icon='close'
                                />
                            </IconButton>
                        </ToolbarGroup>
                    </Toolbar>
                    <iframe
                        style={style.iframe}
                        src={this.props.src}
                    />
                </Paper>
            </div>
        );
    }
}
export default muiThemeable()(withWidth()(IFrame));
