// components/Section.js
import React from 'react';

import FontIcon from 'material-ui/FontIcon';
import IconButton from 'material-ui/IconButton';


class Button extends React.Component {
    render() {
        let t = this.props.theme;
        return (
            <IconButton
                style={{
                    margin: 0,
                    padding: 0,
                    width: t.buttonSize,
                    height: t.buttonSize,
                    transform: this.props.rotated ? 'rotate(90deg)' : 'rotate(0deg)',
                    transition: 'transform 0.5s ease',
                }}
                iconStyle={{
                    fontSize: t.buttonIconSize,
                    margin: 0,
                    padding: 0,
                    color: t.buttonColor
                }}
                onClick={this.props.onClick}
            >
                <FontIcon className='material-icons'>{this.props.icon}</FontIcon>
            </IconButton>
        );
    }
}

class Header extends React.Component {
    render() {
        let t = this.props.theme;

        return (
            <div style={{display: 'flex', alignItems: 'center'}}>
                <div style={{width: t.buttonSize}}>
                    {this.props.toggleable && <Button
                        theme={this.props.theme}
                        rotated={this.props.open}
                        icon={'keyboard_arrow_right'}
                        onClick={this.props.toggleClick}
                    />}
                </div>
                <div style={{
                    flex: 1,
                    display: 'flex',
                    alignItems: 'center',
                    borderBottomStyle: 'solid',
                    minHeight: t.buttonSize,
                    borderBottomWidth: t.underlineWidth,
                    borderBottomColor: t.underlineColor,
                    fontSize: t.fontSize,
                    color: t.textColor
                }}>
                    <div style={{flex: 1}}>
                        {this.props.content}
                    </div>
                    {this.props.icon && <Button
                        theme={this.props.theme}
                        icon={this.props.icon}
                        onClick={this.props.iconClick}
                    />}
                </div>
            </div>
        );
    }
}


export default class Section extends React.Component {
    constructor(props) {
        super(props);
        this.state = {open: ('open' in props) ? props.open : true};
    }

    render() {
        return (
            <div>
                <Header
                    theme={this.props.theme}
                    content={this.props.header}
                    icon={this.props.icon}
                    iconClick={this.props.iconClick}
                    toggleable={!!this.props.children}
                    open={this.state.open}
                    toggleClick={() => this.setState({open: !this.state.open})}
                />

                {this.state.open && <div style={{
                    marginLeft: (this.props.theme.indent || 0),
                }}>
                    {this.props.children}
                </div>}

            </div>
        )
    }
}
