// components/StatusbarWidgets.js
import React from 'react';

import FontIcon from 'material-ui/FontIcon';
import TextField from 'material-ui/TextField';
import CircularProgress from 'material-ui/CircularProgress';
import Slider from 'material-ui/Slider';

import app from 'app';
import MaterialIcon from './MaterialIcon';

class _Widget extends React.Component {
    css() {
        let th = app.theme().gbd.ui.statusbar;
        return {
            group: {
                display: 'inline-flex',
                boxSizing: 'border-box',
                height: '100%',
                padding: '8px 16px',
                alignItems: 'center',

            },

            input: {
                cursor: 'default',
                width: this.props.width || 150,
                fontSize: 12,
                padding: '2px 8px',
                height: '100%',
                boxSizing: 'border-box',
                textAlign: 'left',
                border: 'none',
                outline: 'none',
                backgroundColor: th.input.background,
                color: th.input.color,
            },

            label: {
                padding: '2px 8px',
                fontSize: 12,
                fontWeight: 'bold',
                border: 'none',
                height: '100%',
                boxSizing: 'border-box',
                display: 'inline-flex',
                alignItems: 'center',
                backgroundColor: th.label.background,
                color: th.label.color,
            },

            link: {
                margin: '0 12px',
                fontSize: 12,
                color: th.link.color,
                cursor: 'pointer'
            },

            separator: {
                color: th.separator.color
            },

            sliderContainer: {
                display: 'inline-flex',
                margin: '0 8px',
                padding: 0
            },
            sliderStyle: {},
            sliderSliderStyle: {
                width: 150,
                margin: 0
            }


        }


    }

}


export class Group extends _Widget {
    render() {
        return (
            <div style={this.css().group}>{this.props.children}</div>
        );
    }
}

export class Input extends _Widget {
    render() {
        return (
            <input {...this.props} style={this.css().input}/>
        );
    }
}

export class Label extends _Widget {
    render() {
        return (
            <div style={this.css().label}><span>{this.props.value}</span></div>
        );
    }
}

export class Separator extends _Widget {
    render() {
        return (
            <div style={this.css().separator}><MaterialIcon icon="more_vert"/></div>
        );
    }
}

export class Flex extends _Widget {
    render() {
        return <div style={{flex: 1}}/>
    }
}

export class Progress extends _Widget {
    render() {
        return (
            <Group>
                <CircularProgress size={20}/>
            </Group>
        )
    }
}

export class SmallSlider extends _Widget {
    render() {
        return (
            <div style={this.css().sliderContainer}>
                <Slider
                    {...this.props}
                    style={this.css().sliderStyle}
                    sliderStyle={this.css().sliderSliderStyle}

                />
            </div>
        );
    }


}

export class Link extends _Widget {
    render() {
        return (
            <span
                style={this.css().link}
                href={this.props.target}
            >{this.props.text}</span>
        );
    }
}
