// components/StatusbarWidgets.js
import React from 'react';

import CircularProgress from 'material-ui/CircularProgress';
import Slider from 'material-ui/Slider';

import htmlToReact from 'html-to-react';

import app from 'app';

export class Group extends React.Component {
    render() {
        return (
            <div style={app.theme('gwc.ui.statusbar.group')}>{this.props.children}</div>
        );
    }
}

export class Input extends React.Component {
    render() {
        let style = {
            ...app.theme('gwc.ui.statusbar.input'),
            width: this.props.width || 150
        };

        return (
            <input {...this.props} style={style}/>
        );
    }
}

export class Label extends React.Component {
    render() {
        return (
            <div style={app.theme('gwc.ui.statusbar.label')}><span>{this.props.value}</span></div>
        );
    }
}

export class Separator extends React.Component {
    render() {
        return (
            <div style={app.theme('gwc.ui.statusbar.separator')}>|</div>
        );
    }
}

export class Flex extends React.Component {
    render() {
        return <div style={{flex: 1}}/>
    }
}

export class Progress extends React.Component {
    render() {
        return (
            <Group>
                <CircularProgress size={20}/>
            </Group>
        )
    }
}

export class SmallSlider extends React.Component {
    render() {
        return (
            <div style={app.theme('gwc.ui.statusbar.sliderContainer')}>
                <Slider
                    {...this.props}
                    style={app.theme('gwc.ui.statusbar.sliderStyle')}
                    sliderStyle={app.theme('gwc.ui.statusbar.sliderSliderStyle')}

                />
            </div>
        );
    }


}

export class Link extends React.Component {
    async onClick() {
        let html = await app.http.get(this.props.href);
        let content = new htmlToReact.Parser().parse(html);
        app.perform('dialogShow', {content});
    }

    render() {
        let style = app.theme('gwc.ui.statusbar.link');

        if (this.props.target === 'blank') {
            return (
                <a
                    href={this.props.href}
                    target="_blank"
                    style={style}>{this.props.text}</a>
            );
        }

        if (this.props.target === 'popup') {
            return (
                <span
                    onClick={() => this.onClick()}
                    style={style}>{this.props.text}</span>
            );
        }

        return null;
    }
}
