// components/ToolbarGroup.js
import React from 'react';

import app from '../app';

class ToolbarGroup extends React.Component {
    css() {
        return {
            style: {
                display: 'flex',
                flexDirection: 'row',
                position: 'absolute',
                right: 0,
                boxSizing: 'border-box',
                paddingTop: app.theme().gbd.ui.gutter / 2,
                paddingBottom: app.theme().gbd.ui.gutter / 2,
                paddingRight: 0,
                paddingLeft: app.theme().gbd.ui.gutter / 2,
                borderRadius: '16px 0 0 16px',
                zIndex: 61,
                transform: this.props.visible ? 'translate(0, 0)' : 'translate(100%,0)',
                transition: 'transform 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms',
                top: 0,
                height: app.theme().gbd.ui.toolbar.button.size + app.theme().gbd.ui.gutter,
                background: 'rgba(0, 0, 0, 0.4)'
            }
        }
    }

    render() {
        return (
            <div {...this.css()}>{this.props.children}</div>
        );
    }
}

export default ToolbarGroup;
