// map-util/index.js
import proj4 from 'proj4/lib';
import ol from 'ol-all';

import {get as getMap} from './mapobj';
import config from '../app/config';
import qgis from './qgis';

function initBackground(bg, map) {

    if (bg === 'osm') {
        map.addLayer(new ol.layer.Tile({
            source: new ol.source.OSM(),
        }));
    }

    if (bg === 'webatlas.de-grau') {
        map.addLayer(new ol.layer.Tile({
            source: new ol.source.TileWMS({
                url: 'http://sg.geodatenzentrum.de/wms_webatlasde.light_grau',
                params: {
                    LAYERS: 'webatlasde.light_grau',
                    CRS: config.str('map.proj.client')
                }
            })
        }));
    }


}


async function loadMap() {
    ol.proj.setProj4(proj4);

    let p = config.str('map.proj.server');
    proj4.defs(p, config.str('map.proj.defs.' + p));

    let map = getMap();

    initBackground(config.str('map.background'), map);

    if (config.str('qgis.server')) {
        map = await qgis.load(map);
    }

    map.setView(new ol.View({
        center: ol.proj.fromLonLat(
            config.object('map.center'),
            config.str('map.proj.client')),
        projection: config.str('map.proj.client'),
        zoom: config.number('map.zoom')
    }));

    return map;

}

export default {getMap, loadMap};
