// map-util/index.js
import proj4 from 'proj4/lib';
import ol from 'ol-all';

import {get as getMap} from './mapobj';
import config from '../app/config';
import qgis2 from './qgis2';

function initBackground(bg, map) {

    if (bg === 'osm') {
        map.addLayer(new ol.layer.Tile({
            source: new ol.source.OSM(),
            opacity: 0.3
        }));
    }

    if (bg === 'webatlas.de-grau') {
        map.addLayer(new ol.layer.Tile({
            source: new ol.source.TileWMS({
                url: 'http://sg.geodatenzentrum.de/wms_webatlasde.light_grau',
                params: {
                    LAYERS: 'webatlasde.light_grau',
                    CRS: config.str('map.crs.client')
                }
            })
        }));
    }
}

function viewOpts() {
    return {
        center: config.object('map.center'),
        projection: config.str('map.crs.client'),
        zoom: config.number('map.zoom.init'),
        minZoom: config.number('map.zoom.min'),
        maxZoom: config.number('map.zoom.max'),
    };
}


async function loadMap() {
    ol.proj.setProj4(proj4);

    let p = config.str('map.crs.server');
    proj4.defs(p, config.str('map.crs.defs.' + p));

    let map = getMap();

    initBackground(config.str('map.background'), map);

    let vo = viewOpts();

    if (config.str('qgis2.server')) {
        let qp = await qgis2.loadWMS(map);
        map.addLayer(qp.rootLayer);
        vo.extent = qp.extent;
    }


    map.setView(new ol.View(vo));
    return map;

}

export default {getMap, loadMap, qgis2};
