// map-util/loader.js
/// map loader

import proj4 from 'proj4/lib';
import ol from 'ol-all';

import config from '../app/config';


function backgroundLayer() {
    let bg = config.str('map.background');

    if (bg === 'osm') {
        return new ol.layer.Tile({
            source: new ol.source.OSM(),
            name: 'OSM',
            type: 'BackgroundLayer',
            opacity: 0.3
        });
    }

    if (bg === 'webatlas.de-grau') {
        return new ol.layer.Tile({
            source: new ol.source.TileWMS({
                name: 'Webatlas',
                title: 'Webatlas',
                type: 'BackgroundLayer',
                url: 'http://sg.geodatenzentrum.de/wms_webatlasde.light_grau',
                params: {
                    LAYERS: 'webatlasde.light_grau',
                    CRS: config.str('map.crs.client')
                }
            })
        });
    }
}

function viewOpts() {
    return {
        center: config.object('map.center'),
        projection: config.str('map.crs.client'),
        zoom: config.number('map.zoom.init'),
        minZoom: config.number('map.zoom.min'),
        maxZoom: config.number('map.zoom.max'),
    };
}


async function load(map) {
    ol.proj.setProj4(proj4);

    let p = config.str('map.crs.server');
    proj4.defs(p, config.str('map.crs.defs.' + p));

    let bg = backgroundLayer();
    if(bg) {
        map.addLayer(new ol.layer.Group({
            name: 'Backgrounds',
            zIndex: 0,
            type: 'BackgroundRoot',
            layers: [bg]
        }));
    }

    map.setView(new ol.View(viewOpts()));
    return map;

}

export default {load};
