// map-util/loader.js
/// map loader

import ol from 'ol-all';

import config from '../app/config';


function backgroundLayer() {
    let bg = config.str('map.background');

    if (bg === 'osm') {
        return new ol.layer.Tile({
            source: new ol.source.OSM(),
            name: 'OSM',
            kind: 'BackgroundLayer',
            opacity: 0.3
        });
    }

    if (bg === 'webatlas.de-grau') {
        return new ol.layer.Tile({
            source: new ol.source.TileWMS({
                name: 'Webatlas',
                title: 'Webatlas',
                kind: 'BackgroundLayer',
                url: 'http://sg.geodatenzentrum.de/wms_webatlasde.light_grau',
                params: {
                    LAYERS: 'webatlasde.light_grau',
                    CRS: config.str('map.crs.client')
                }
            })
        });
    }
}

async function load(map) {
    let bg = backgroundLayer();
    if(bg) {
        map.attachLayer('background', bg);
    }

    return map;

}

export default {load};
