// map-util/mapobj.js
import ol from 'ol-all';
import pubsub from '../app/pubsub';

class TMap extends ol.Map {
    constructor() {
        super({
            controls: [],
            interactions: []
        });
        this.currentMode = null;
    }

    setMode(name, cursor = null, interactions = null) {
        if (name === this.currentMode)
            return;

        if (name === 'default') {
            cursor = 'default';
            interactions = [
                new ol.interaction.DragPan(),
                new ol.interaction.PinchZoom(),
                new ol.interaction.MouseWheelZoom(),
            ]
        }


        if (this.currentMode) {
            pubsub.publish('map.mode.end', this.currentMode);
        }

        this.getInteractions().clear();
        for (let i of interactions)
            this.addInteraction(i);

        this.getTargetElement().style.cursor = cursor;
        this.currentMode = name;
        pubsub.publish('map.mode.start', this.currentMode);
    }

    defaultMode() {
        this.setMode('default');
    }
}

let _map = null;

/**
 *
 * @returns {TMap}
 */
export function get() {
    if (_map)
        return _map;

    _map = new TMap();
    if (window)
        window._map = _map;

    return _map;
}
