// map-util/wms/request.js
import _ from 'lodash';
import app from 'app';

async function get(url, verb, params) {
    return await app.http.get(url, {
        service: 'WMS',
        version: '1.3',
        request: verb,
        ...params
    });
}


async function getXML(url, verb, params) {
    let r = await get(url, verb, params);
    let p = new DOMParser();
    return p.parseFromString(r, 'application/xml');
}


function toJs(el) {

    let noNs = s => s.split(':').pop();

    let p = {'@attributes': {}, '@text': []};

    if (el.attributes) {
        for (let i = 0; i < el.attributes.length; i++) {
            let attr = el.attributes.item(i);
            p['@attributes'][noNs(attr.nodeName)] = attr.nodeValue;
        }
    }

    if (el.childNodes) {
        for (let i = 0; i < el.childNodes.length; i++) {
            let cc = el.childNodes.item(i);

            if (cc.nodeType === 3) {
                let v = String(cc.nodeValue).trim();
                if (v)
                    p['@text'].push(v);
                continue;
            }

            if (cc.nodeType === 1) {
                let name = noNs(cc.nodeName);
                p[name] = p[name] || [];
                p[name].push(toJs(cc));

            }
        }
    }

    return p;
}

async function getJs(url, verb, params) {
    let doc = await getXML(url, verb, params);
    return toJs(doc.firstChild);

}

export default {
    get,
    getXML,
    getJs
}
